src/github.com/getlantern/flashlight/client/handler.go

```diff
@@ -69,7 +73,11 @@ func (client *Client) intercept(resp http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	defer connOut.Close()
+	defer func() {
+		if err := connOut.Close(); err != nil {
+			log.Debugf("Error closing the out connection: %s", err)
+		}
+	}()
```

Comment:
I think that's right yeah but I would just add the closing method to intercept, so something like:

```
 44     var err error
 45
 46     // Hijack underlying connection.
 47     var clientConn net.Conn
 48
 49     var connOut net.Conn
 50
 51     // Make sure of closing connections only once
 52     var closeOnce sync.Once
 53
 54     // Force closing if EOF at the request half or error encountered.
 55     // A bit arbitrary, but it's rather rare now to use half closing
 56     // as a way to notify server. Most application closes both connections
 57     // after completed send / receive so that won't cause problem.
 58     closeConns := func() {
 59         if clientConn != nil && err := clientConn.Close(); err != nil {
 60             log.Debugf("Error closing the out connection: %s", err)
 61         }
 62         if connOut != nil && err := connOut.Close(); err != nil {
 63             log.Debugf("Error closing the client connection: %s", err)
 64         }
 65     }
 66
 67     defer closeOnce.Do(closeConns)
 68
 69     if clientConn, _, err = resp.(http.Hijacker).Hijack(); err != nil {
 70         respondBadGateway(resp, fmt.Sprintf("Unable to hijack connection: %s", err))
 71         return
 72     }
 73
 74     addr := hostIncludingPort(req, 443)
 75     // Establish outbound connection.
 76     d := func(network, addr string) (net.Conn, error) {
 77         return client.getBalancer().DialQOS("tcp", addr, client.targetQOS(req))
 78     }
 79
 80     if runtime.GOOS == "android" || client.ProxyAll {
 81         connOut, err = d("tcp", addr)
 82     } else {
 83         connOut, err = detour.Dialer(d)("tcp", addr)
 84     }
 85
 86     if err != nil {
 87         respondBadGateway(clientConn, fmt.Sprintf("Unable to handle CONNECT request: %s", err))
 88         return
 89     }
 90
 91     // Pipe data between the client and the proxy.
 92     pipeData(clientConn, connOut, req, closeConns)
```